File: fushi/test/utils/misc/popup_dict_css_atrule_scope_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

/// TODO-812: 明鏡国語辞典 第三版 (and any Yomitan dictionary whose `styles.css`
/// contains at-rules) must render correctly in the lookup popup. The legacy
/// `constructDictCss` in `assets/popup/dict-media.js` is a hand-written CSS
/// scoper that prefixed every comma-separated selector with
/// `[data-dictionary="X"]`. For an `@media (max-width: 500px) { ... }` block it
/// treated `@media (max-width: 500px)` as a selector list, emitting the illegal
/// `[data-dictionary="X"] @media (...) { ... }`, so the whole @media block was
/// dropped by the browser — the meikyo responsive image sizing vanished.
///
/// 根因：`dict-media.js` `constructDictCss` 把 at-rule 前言当成选择器逐个加前缀。
/// 修复：识别 at-rule —— 条件组 (@media/@supports/@container/@layer/@scope) 前言
/// 原样保留、对内部规则递归加前缀；其它 at-rule (@font-face/@keyframes/@page) 整块
/// 透传；语句型 at-rule (@import/@charset/...) 原样输出。
///
/// 两层守护：
/// ① 行为级——用 Node 真执行 `constructDictCss` 喂明鏡真实 styles.css 片段，断言
///    @media 前言不被前缀污染、内部规则仍作用域化、普通选择器不受影响、
///    @keyframes/@font-face 体不被前缀。无 node 时 skip。
/// ② 源码级——静态扫描 dict-media.js，保证 at-rule 识别分支在位（CI 无 node 也守住）。
void main() {
  test(
    'constructDictCss scopes at-rules correctly (executes via node)',
    () async {
      final String? nodeExe = _resolveNode();
      if (nodeExe == null) {
        markTestSkipped(
            'node not found on PATH; skipping JS behavior execution');
        return;
      }

      final File jsTest =
          File('test/utils/misc/popup_dict_css_atrule_scope_test.js');
      expect(
        jsTest.existsSync(),
        isTrue,
        reason: 'behavior harness ${jsTest.path} must exist',
      );

      final ProcessResult result = await Process.run(
        nodeExe,
        <String>[jsTest.path],
        workingDirectory: Directory.current.path,
      );

      expect(
        result.exitCode,
        0,
        reason: 'dict CSS at-rule scoping JS behavior test failed.\n'
            'stdout:\n${result.stdout}\nstderr:\n${result.stderr}',
      );
      expect(
        result.stdout.toString(),
        contains('all assertions passed'),
        reason: 'behavior harness must reach its success marker',
      );
    },
  );

  test('dict-media.js constructDictCss recognizes at-rules', () {
    final String js = File('assets/popup/dict-media.js').readAsStringSync();

    // The at-rule detection branch must exist and not be a no-op.
    expect(
      js,
      contains(r'const atRuleMatch = selectorPrelude.match(/^@([a-z-]+)/i);'),
      reason: 'constructDictCss must detect at-rule preludes',
    );
    // Conditional-group at-rules must recurse to keep inner rules scoped.
    final int atBranch = js.indexOf('const atRuleMatch');
    expect(atBranch, greaterThanOrEqualTo(0));
    final int isCond = js.indexOf('isConditionalGroup', atBranch);
    expect(
      isCond,
      greaterThan(atBranch),
      reason: 'at-rule branch must split conditional groups from the rest',
    );
    final int recurse = js.indexOf(
        'constructDictCss(atBlockContent, dictName, scopePrefix)', atBranch);
    expect(
      recurse,
      greaterThan(atBranch),
      reason: 'conditional-group at-rules must recurse to scope inner rules',
    );
    // @media must be listed as a conditional group.
    final int mediaListed = js.indexOf("atName === 'media'", atBranch);
    expect(
      mediaListed,
      greaterThan(atBranch),
      reason: '@media must be treated as a conditional group',
    );
  });
}

/// Resolve a usable `node` executable, returning null when none is on PATH.
String? _resolveNode() {
  final List<String> candidates =
      Platform.isWindows ? <String>['node.exe', 'node'] : <String>['node'];
  for (final String name in candidates) {
    try {
      final ProcessResult probe = Process.runSync(name, <String>['--version']);
      if (probe.exitCode == 0) {
        return name;
      }
    } on ProcessException {
      // Not found; try next candidate.
    }
  }
  return null;
}

File: fushi/test/utils/misc/popup_dict_css_atrule_scope_test.js
const assert = require('assert');
const fs = require('fs');
const path = require('path');
const vm = require('vm');

// TODO-812: 明鏡国語辞典 第三版 ships a styles.css that contains an `@media`
// at-rule. The legacy `constructDictCss` (a hand-written CSS scoper) treated the
// whole text before the first `{` as a selector list, so `@media (max-width:
// 500px)` got the `[data-dictionary="X"]` prefix glued in front, producing the
// ILLEGAL `[data-dictionary="X"] @media (...) { ... }` — the browser drops the
// entire @media block. This harness executes the real `constructDictCss` and
// asserts at-rules are scoped correctly while ordinary selectors keep their
// prefix. Mirrors the Hoshi-Reader-Android rendering baseline.

const dictMediaPath = path.resolve(__dirname, '../../../assets/popup/dict-media.js');

function load() {
  const context = { console };
  context.globalThis = context;
  vm.runInNewContext(fs.readFileSync(dictMediaPath, 'utf8'), context, {
    filename: dictMediaPath,
  });
  return context;
}

const DICT = '明鏡国語辞典 第三版';
const PREFIX = `.yomitan-glossary [data-dictionary="${DICT}"]`;

function scope(css) {
  return load().constructDictCss(css, DICT, PREFIX);
}

// Real fragment lifted verbatim from the meikyo styles.css.
const MEIKYO_MEDIA = `
table[data-sc-class="img"] .gloss-image-link {
    max-width: 70% !important;
}
@media (max-width: 500px) {
	table[data-sc-class="img"] .gloss-image-link {
		max-width: 65% !important;
	}
}
div[data-sc-head2] {
	margin-block-start: 0.25em;
}`;

function testMediaPreludeNotPrefixed() {
  const out = scope(MEIKYO_MEDIA);
  // The @media prelude must stay raw, never glued behind the scope prefix.
  assert.ok(
    !/\[data-dictionary="[^"]*"\]\s*@media/.test(out),
    '@media prelude was illegally prefixed:\n' + out,
  );
  assert.ok(
    /@media\s*\(max-width:\s*500px\)\s*\{/.test(out),
    '@media prelude missing or mangled:\n' + out,
  );
}

function testMediaInnerRuleStillScoped() {
  const out = scope(MEIKYO_MEDIA);
  // Inside @media, the nested style rule MUST still receive the scope prefix.
  const mediaStart = out.indexOf('@media');
  const inner = out.slice(mediaStart);
  assert.ok(
    inner.includes(`${PREFIX} table[data-sc-class="img"] .gloss-image-link`),
    'inner @media rule lost its scope prefix:\n' + out,
  );
}

function testOrdinaryRulesUnaffected() {
  const out = scope(MEIKYO_MEDIA);
  assert.ok(
    out.includes(`${PREFIX} div[data-sc-head2]`),
    'ordinary selector lost its prefix (regression):\n' + out,
  );
  assert.ok(
    out.includes(`${PREFIX} table[data-sc-class="img"] .gloss-image-link`),
    'ordinary selector before @media lost its prefix (regression):\n' + out,
  );
}

function testKeyframesBodyNeverPrefixed() {
  const out = scope('@keyframes spin { 0% { opacity: 0; } 100% { opacity: 1; } }');
  assert.ok(/@keyframes spin\s*\{/.test(out), '@keyframes prelude mangled:\n' + out);
  assert.ok(
    !/\[data-dictionary="[^"]*"\]\s*0%/.test(out),
    '@keyframes selector (0%) was illegally prefixed:\n' + out,
  );
  assert.ok(out.includes('0% { opacity: 0; }'), '@keyframes body altered:\n' + out);
}

function testFontFaceBodyNeverPrefixed() {
  const out = scope('@font-face { font-family: X; src: url(a.woff); }');
  assert.ok(
    !/\[data-dictionary="[^"]*"\]\s*@font-face/.test(out),
    '@font-face prelude was illegally prefixed:\n' + out,
  );
  assert.ok(out.includes('font-family: X;'), '@font-face body altered:\n' + out);
}

function testStatementAtRulePassthrough() {
  const out = scope('@import "x.css"; span[data-sc-b] { color: red; }');
  assert.ok(out.includes('@import "x.css";'), '@import statement not preserved:\n' + out);
  assert.ok(
    !/\[data-dictionary="[^"]*"\]\s*@import/.test(out),
    '@import was illegally prefixed:\n' + out,
  );
  assert.ok(
    out.includes(`${PREFIX} span[data-sc-b]`),
    'rule after @import statement lost its prefix:\n' + out,
  );
}

function testCjkDataKeyExpansionMatchesBaseline() {
  // Guard the data-key → attribute expansion contract shared with the
  // Hoshi-Reader-Android baseline: CJK-leading keys drop the hyphen
  // (`data-sc` + key) while ASCII keys keep it (`data-sc-<kebab>`). This is the
  // intentional daijisen/meikyo behavior; both apps agree. Replicated from
  // popup.js so the guard is self-contained.
  function toKebabCase(str) {
    return str.replace(/([A-Z])/g, (_, c, i) => (i ? '-' : '') + c.toLowerCase());
  }
  function attrName(k) {
    const isCJK = /^[　-鿿豈-﫿]/.test(k);
    return `data-sc${isCJK ? '' : '-'}${toKebabCase(k)}`;
  }
  assert.strictEqual(attrName('headword'), 'data-sc-headword');
  assert.strictEqual(attrName('class'), 'data-sc-class');
  assert.strictEqual(attrName('カナ'), 'data-scカナ');
  assert.strictEqual(attrName('表記'), 'data-sc表記');
}

testMediaPreludeNotPrefixed();
testMediaInnerRuleStillScoped();
testOrdinaryRulesUnaffected();
testKeyframesBodyNeverPrefixed();
testFontFaceBodyNeverPrefixed();
testStatementAtRulePassthrough();
testCjkDataKeyExpansionMatchesBaseline();

console.log('all assertions passed');
